### PR TITLE
PP-6095 remove charge entity from refund entity

### PR DESCRIPTION
**Pull request by SandorArpa**
## WHAT YOU DID
- de-couple ChargeEntity from RefundEntity
- re-factor methods to pass in Charge/ChargeEntity
- re-factor tests accordingly

### CODE REVIEW

File: src/main/java/uk/gov/pay/connector/events/EmittedEventsBackfillService.java [hunk 1]
```diff
@@ -80,12 +82,17 @@ public void backfillNotEmittedEvents() {
...
         try {
-            Optional<ChargeEntity> charge;
+            Optional<ChargeEntity> charge = Optional.empty();
 
             if (ResourceType.valueOf(event.getResourceType().toUpperCase()).equals(ResourceType.PAYMENT)) {
-                charge = Optional.of(chargeService.findChargeById(event.getResourceExternalId()));
+                charge = Optional.of(chargeService.findChargeByExternalId(event.getResourceExternalId()));
             } else {
-                charge = refundDao.findByExternalId(event.getResourceExternalId()).map(RefundEntity::getChargeEntity);
+                Optional<RefundEntity> refundEntity = refundDao.findRefundsByChargeExternalId(event.getResourceExternalId())
```
Comment (kbottla) [inline src/main/java/uk/gov/pay/connector/events/EmittedEventsBackfillService.java]: 1. This should get refundEntity based on `refundExternalId` and then a charge to be retrieved to emit ALL events (payment and refunds).
EmittedEventEntity stores refundExternalId for refund events

2. In the first instance (in this PR) we can assume charge is not expunged yet so we can get from DB. We need to look separately for how we emit refund events / refund availability for expunged charges. (cc @sfount )

File: src/test/java/uk/gov/pay/connector/util/DatabaseTestHelper.java [hunk 2]
```diff
@@ -173,14 +173,15 @@ public void addRefundHistory(long id, String externalId, String reference, long
...
         );
     }
 
-    public void addRefundHistory(long id, String externalId, String reference, long amount, String status, Long chargeId, ZonedDateTime createdDate, ZonedDateTime historyStartDate, String submittedByUserExternalId) {
+    public void addRefundHistory(long id, String externalId, String reference, long amount, String status, Long chargeId, ZonedDateTime createdDate, ZonedDateTime historyStartDate, String submittedByUserExternalId, String chargeExternalId) {
         jdbi.withHandle(handle ->
                 handle
-                        .createUpdate("INSERT INTO refunds_history(id, external_id, reference, amount, status, charge_id, created_date, history_start_date, user_external_id) VALUES (:id, :external_id, :reference, :amount, :status, :charge_id, :created_date, :history_start_date, :user_external_id)")
+                        .createUpdate("INSERT INTO refunds_history(id, external_id, reference, amount, status, charge_id, created_date, history_start_date, user_external_id, charge_external_id) VALUES (:id, :external_id, :reference, :amount, :status, :charge_id, :created_date, :history_start_date, :user_external_id, :charge_external_id)")
```
Comment (kbottla) [inline src/test/java/uk/gov/pay/connector/util/DatabaseTestHelper.java:185]: same as above for charge_id

File: src/test/java/uk/gov/pay/connector/util/DatabaseTestHelper.java [hunk 3]
```diff
@@ -157,10 +157,10 @@ public int addRefund(String externalId, String reference, long amount, RefundSta
...
+    public void addRefundHistory(long id, String externalId, String reference, long amount, String status, Long chargeId, ZonedDateTime createdDate, ZonedDateTime historyStartDate, ZonedDateTime historyEndDate, String submittedByUserExternalId, String userEmail, String chargeExternalId) {
         jdbi.withHandle(handle ->
                 handle
-                        .createUpdate("INSERT INTO refunds_history(id, external_id, reference, amount, status, charge_id, created_date, history_start_date, history_end_date, user_external_id, user_email) VALUES (:id, :external_id, :reference, :amount, :status, :charge_id, :created_date, :history_start_date, :history_end_date, :user_external_id, :user_email)")
+                        .createUpdate("INSERT INTO refunds_history(id, external_id, reference, amount, status, charge_id, created_date, history_start_date, history_end_date, user_external_id, user_email, charge_external_id) VALUES (:id, :external_id, :reference, :amount, :status, :charge_id, :created_date, :history_start_date, :history_end_date, :user_external_id, :user_email, :charge_external_id)")
```
Comment (kbottla) [inline src/test/java/uk/gov/pay/connector/util/DatabaseTestHelper.java:164]: `charge_id` can be removed here too (same as refunds https://github.com/alphagov/pay-connector/pull/2091/files#diff-59b289c8182a4fde50257daf15bbaddcR282) - fine as separate PR

File: src/test/java/uk/gov/pay/connector/usernotification/service/UserNotificationServiceTest.java [hunk 4]
```diff
@@ -123,7 +123,7 @@ public void shouldSendPaymentConfirmationEmailIfEmailNotifyIsEnabled() throws Ex
                 personalisation,
                 null)).thenReturn(mockNotificationCreatedResponse);
 
-        Optional<String> maybeNotificationId = userNotificationService.sendPaymentConfirmedEmail(chargeEntity).get(1000, TimeUnit.SECONDS);
+        Optional<String> maybeNotificationId = userNotificationService.sendPaymentConfirmedEmail(chargeEntity, chargeEntity.getGatewayAccount()).get(1000, TimeUnit.SECONDS);
```
Comment (kbottla) [inline src/test/java/uk/gov/pay/connector/usernotification/service/UserNotificationServiceTest.java:131]: Not relevant to this PR but UserNotifications should not rely on chargeEntity. To be addressed in PP-6066 or separate story (cc @sfount )

File: src/main/java/uk/gov/pay/connector/paymentprocessor/service/CardCaptureService.java [hunk 5]
```diff
@@ -82,7 +82,7 @@ public ChargeEntity markChargeAsEligibleForCapture(String externalId) {
 
         if (!charge.isDelayedCapture()) {
             addChargeToCaptureQueue(charge);
-            userNotificationService.sendPaymentConfirmedEmail(charge);
+            userNotificationService.sendPaymentConfirmedEmail(charge, charge.getGatewayAccount());
```
Comment (kbottla) [inline src/main/java/uk/gov/pay/connector/paymentprocessor/service/CardCaptureService.java:85]: Do we need to change `sendPaymentConfirmedEmail` signature as it is only used for inflight payments ?

File: src/main/java/uk/gov/pay/connector/refund/dao/RefundDao.java [hunk 6]
```diff
@@ -59,31 +59,31 @@ public RefundDao(final Provider<EntityManager> entityManager) {
...
 
-    public List<RefundHistory> searchAllHistoryByChargeId(Long chargeId) {
+    public List<RefundHistory> searchAllHistoryByChargeExternalId(String chargeExternalId) {
         String query = "SELECT r.id, r.external_id, r.amount, r.status, charge_id, r.created_date, r.version, " +
                 "r.reference, history_start_date, history_end_date, user_external_id, r.gateway_transaction_id, " +
                 "c.external_id AS charge_external_id, c.gateway_account_id, r.user_email " +
                 " FROM refunds_history r , charges c" +
-                " WHERE charge_id = ?1" +
-                " and r.charge_id = c.id";
+                " WHERE charge_external_id = ?1" +
```
Comment (kbottla) [inline src/main/java/uk/gov/pay/connector/refund/dao/RefundDao.java]: Can we remove join to `Charges` here too ? and not have gateway_account_id in query. If so can be a separate PR

File: src/main/java/uk/gov/pay/connector/refund/model/domain/RefundEntity.java [hunk 7]
```diff
@@ -93,18 +88,20 @@
     @Column(name = "charge_external_id")
     private String chargeExternalId;
 
+    @Column(name = "charge_id")
```
Comment (kbottla) [inline src/main/java/uk/gov/pay/connector/refund/model/domain/RefundEntity.java:91]: Do we need `chargeId` any more ? We might drop this column too since we have chargeExternalId

File: src/main/java/uk/gov/pay/connector/refund/service/ChargeRefundService.java [hunk 8]
```diff
@@ -61,45 +62,43 @@ public ChargeRefundService(ChargeService chargeService, RefundDao refundDao, Gat
     public ChargeRefundResponse doRefund(Long accountId, String chargeExternalId, RefundRequest refundRequest) {
         GatewayAccountEntity gatewayAccountEntity = gatewayAccountDao.findById(accountId).orElseThrow(
                 () -> new GatewayAccountNotFoundException(accountId));
-        RefundEntity refundEntity = createRefund(accountId, chargeExternalId, refundRequest);
+        ChargeEntity chargeEntity = chargeDao.findByExternalId(chargeExternalId)
```
Comment (kbottla) [inline src/main/java/uk/gov/pay/connector/refund/service/ChargeRefundService.java]: Could we still use chargeService to find ChargeEntity ? This https://github.com/alphagov/pay-connector/pull/2085 ensures an object returned either from connector or ledger

File: src/main/java/uk/gov/pay/connector/refund/dao/RefundDao.java [hunk 9]
```diff
@@ -59,31 +59,31 @@ public RefundDao(final Provider<EntityManager> entityManager) {
 
     }
 
-    public List<RefundHistory> searchHistoryByChargeId(Long chargeId) {
+    public List<RefundHistory> searchHistoryByChargeExternalId(String chargeExternalId) {
 
         String query = "SELECT id, external_id, amount, status, charge_id, created_date, version, reference, " +
                 "history_start_date, history_end_date, user_external_id, gateway_transaction_id, user_email, charge_external_id " +
                 "FROM refunds_history r " +
-                "WHERE charge_id = ?1 AND status != ?2";
```
Comment (sfount) [inline src/main/java/uk/gov/pay/connector/refund/dao/RefundDao.java:67]: We currently only have an index on the charge id in the refund history table, I wonder how important this query is but if we are going to support searching by external ID we'll need to add one for that field too

```
Indexes:
    "idx_refunds_history_charge_id" btree (charge_id)
```

File: src/main/java/uk/gov/pay/connector/refund/dao/RefundDao.java [hunk 10]
```diff
@@ -59,31 +59,31 @@ public RefundDao(final Provider<EntityManager> entityManager) {
...
+                .setParameter(1, chargeExternalId)
                 .setParameter(2, RefundStatus.CREATED.getValue())
                 .getResultList();
     }
 
-    public List<RefundHistory> searchAllHistoryByChargeId(Long chargeId) {
+    public List<RefundHistory> searchAllHistoryByChargeExternalId(String chargeExternalId) {
         String query = "SELECT r.id, r.external_id, r.amount, r.status, charge_id, r.created_date, r.version, " +
                 "r.reference, history_start_date, history_end_date, user_external_id, r.gateway_transaction_id, " +
                 "c.external_id AS charge_external_id, c.gateway_account_id, r.user_email " +
```
Comment (sfount) [inline src/main/java/uk/gov/pay/connector/refund/dao/RefundDao.java]: Potentially not part of this commit/ PR but we won't be able to rely on this join for the refund history table going forward

File: src/main/java/uk/gov/pay/connector/refund/service/ChargeRefundService.java [hunk 11]
```diff
@@ -161,11 +160,10 @@ private RefundStatus determineRefundStatus(GatewayRefundResponse gatewayRefundRe
...
     @SuppressWarnings("WeakerAccess")
-    public RefundEntity createRefundEntity(RefundRequest refundRequest, Charge charge) {
-        // @TODO(sfount) PP-6095 remove charge entity from refund entity - we can then rely solely on the Charge POJO
-        ChargeEntity chargeEntity = chargeService.findChargeById(charge.getExternalId());
-        RefundEntity refundEntity = new RefundEntity(chargeEntity, refundRequest.getAmount(),
-                refundRequest.getUserExternalId(), refundRequest.getUserEmail());
+    public RefundEntity createRefundEntity(RefundRequest refundRequest, ChargeEntity charge) {
+        RefundEntity refundEntity = new RefundEntity(refundRequest.getAmount(),
+                refundRequest.getUserExternalId(), refundRequest.getUserEmail(), charge.getExternalId());
+        refundEntity.setChargeId(charge.getId());
```
Comment (kbottla) [inline src/main/java/uk/gov/pay/connector/refund/service/ChargeRefundService.java]: Might not need this `        refundEntity.setChargeId(charge.getId());
`

File: src/main/java/uk/gov/pay/connector/usernotification/service/UserNotificationService.java [hunk 12]
```diff
@@ -63,22 +63,23 @@ public UserNotificationService(NotifyClientFactory notifyClientFactory, Connecto
...
-    public Future<Optional<String>> sendRefundIssuedEmail(RefundEntity refundEntity) {
-        return sendEmail(EmailNotificationType.REFUND_ISSUED, refundEntity.getChargeEntity(), buildRefundEmailPersonalisationFrom(refundEntity));
+    public Future<Optional<String>> sendRefundIssuedEmail(RefundEntity refundEntity, ChargeEntity chargeEntity, GatewayAccountEntity gatewayAccountEntity) {
+        return sendEmail(EmailNotificationType.REFUND_ISSUED, chargeEntity, gatewayAccountEntity,
+                buildRefundEmailPersonalisationFrom(chargeEntity, refundEntity));
     }
 
-    public Future<Optional<String>> sendPaymentConfirmedEmail(ChargeEntity chargeEntity) {
-        return sendEmail(EmailNotificationType.PAYMENT_CONFIRMED, chargeEntity, buildConfirmationEmailPersonalisationFrom(chargeEntity));
+    public Future<Optional<String>> sendPaymentConfirmedEmail(ChargeEntity chargeEntity, GatewayAccountEntity gatewayAccountEntity) {
```
Comment (kbottla) [inline src/main/java/uk/gov/pay/connector/usernotification/service/UserNotificationService.java:72]: Do we need to update this method ? As this is only used for inflight payments, we should always have gatewayAccount on chargeEntity.

File: src/main/java/uk/gov/pay/connector/refund/model/domain/RefundEntity.java [hunk 13]
```diff
@@ -47,7 +47,6 @@
                         @ColumnResult(name = "user_external_id", type = String.class),
                         @ColumnResult(name = "gateway_transaction_id", type = String.class),
                         @ColumnResult(name = "charge_external_id", type = String.class),
-                        @ColumnResult(name = "gateway_account_id", type = Long.class),
```
Comment (sfount) [inline src/main/java/uk/gov/pay/connector/refund/model/domain/RefundEntity.java:50]: This will likely change how refunds and refund history is populated -- we should probably make sure we cleanup the SQL to reflect that this will no longer be serialised/ deserialised

**Review by kbottla — COMMENTED** [review]
Halfway through and will commence to review the rest(tests) shortly .

Mostly looks good. Added a few minor comments and blocking one for `EmittedEventsBackfillService` where we need to find refunds by refundExternalId

File: src/main/java/uk/gov/pay/connector/refund/model/domain/RefundHistory.java [hunk 14]
```diff
@@ -19,24 +18,14 @@
...
-                         String userExternalId, String gatewayTransactionId, String chargeExternalId, Long gatewayAccountId,
+                         String userExternalId, String gatewayTransactionId, String chargeExternalId,
                          String userEmail) {
         super();
         setId(id);
         setExternalId(externalId);
         setAmount(amount);
         setStatus(status);
-
-        GatewayAccountEntity gatewayAccount = new GatewayAccountEntity();
```
Comment (sfount) [inline src/main/java/uk/gov/pay/connector/refund/model/domain/RefundHistory.java:30]: 👍 assuming refund histories should never need to create public API reference links. This is a good cleanup of these processes

File: src/main/java/uk/gov/pay/connector/refund/resource/ChargeRefundsResource.java [hunk 15]
```diff
@@ -91,11 +91,11 @@ public Response getRefunds(@PathParam("accountId") Long accountId, @PathParam("c
                 .orElse(responseWithChargeNotFound(chargeId));
     }
 
-    private Response getRefundResponse(ChargeEntity chargeEntity, String refundId, UriInfo uriInfo) {
+    private Response getRefundResponse(ChargeEntity chargeEntity, String refundId, Long accountId, UriInfo uriInfo) {
```
Comment (sfount) [inline src/main/java/uk/gov/pay/connector/refund/resource/ChargeRefundsResource.java:94]: Not part of this PR but reminder to myself, these methods will have to updated to use the `Charge` model introduced in previous PRs which also exposes the `getExternalId` but won't rely on the database

Comment (sfount) [reply src/main/java/uk/gov/pay/connector/refund/resource/ChargeRefundsResource.java:94]: In particular this account ID will have to be provided from the top https://github.com/alphagov/pay-connector/blob/master/src/main/java/uk/gov/pay/connector/refund/model/RefundResponse.java#L19

File: src/main/java/uk/gov/pay/connector/refund/service/ChargeRefundService.java [hunk 16]
```diff
@@ -61,45 +62,43 @@ public ChargeRefundService(ChargeService chargeService, RefundDao refundDao, Gat
...
     @Transactional
     @SuppressWarnings("WeakerAccess")
-    public RefundEntity createRefund(Long accountId, String chargeExternalId, RefundRequest refundRequest) {
-        GatewayAccountEntity gatewayAccountEntity = gatewayAccountDao.findById(accountId).orElseThrow(
-                () -> new GatewayAccountNotFoundException(accountId));
-        return chargeService.findCharge(chargeExternalId, gatewayAccountEntity.getId()).map(charge -> {
-            List<RefundEntity> refundEntityList = refundDao.findRefundsByChargeExternalId(charge.getExternalId());
-            long availableAmount = validateRefundAndGetAvailableAmount(charge, gatewayAccountEntity, refundRequest, refundEntityList);
-            RefundEntity refundEntity = createRefundEntity(refundRequest, charge);
+    public RefundEntity createRefund(GatewayAccountEntity gatewayAccountEntity, ChargeEntity chargeEntity, RefundRequest refundRequest) {
```
Comment (sfount) [inline src/main/java/uk/gov/pay/connector/refund/service/ChargeRefundService.java]: Using a charge entity here directly unpicks the PR to make this agnostic (merged in https://github.com/alphagov/pay-connector/pull/2082) -- it would be good to make this use the `Charge` where possible and potentially rethink the interface here. Will see if I can make a suggestion for how it might look

File: src/main/java/uk/gov/pay/connector/refund/service/ChargeRefundService.java [hunk 17]
```diff
@@ -161,11 +160,10 @@ private RefundStatus determineRefundStatus(GatewayRefundResponse gatewayRefundRe
 
     @Transactional
     @SuppressWarnings("WeakerAccess")
-    public RefundEntity createRefundEntity(RefundRequest refundRequest, Charge charge) {
-        // @TODO(sfount) PP-6095 remove charge entity from refund entity - we can then rely solely on the Charge POJO
-        ChargeEntity chargeEntity = chargeService.findChargeById(charge.getExternalId());
-        RefundEntity refundEntity = new RefundEntity(chargeEntity, refundRequest.getAmount(),
-                refundRequest.getUserExternalId(), refundRequest.getUserEmail());
+    public RefundEntity createRefundEntity(RefundRequest refundRequest, ChargeEntity charge) {
```
Comment (sfount) [inline src/main/java/uk/gov/pay/connector/refund/service/ChargeRefundService.java]: As above, retied everything back to the `ChargeEntity`

**Review by sfount — COMMENTED** [review]
Code is looking good, I've got some comments on making sure this lines up with the work to stop relying on the `ChargeEntity`, it looks like it's been introduced to a lot of the methods during the refund process. Will further clarify comments if needed

File: src/main/java/uk/gov/pay/connector/refund/model/domain/RefundEntity.java (hunk 7 above)
Comment (SandorArpa) [reply src/main/java/uk/gov/pay/connector/refund/model/domain/RefundEntity.java:91]: We do until the foreign key constraint is removed

File: src/main/java/uk/gov/pay/connector/refund/dao/RefundDao.java (hunk 10 above)
Comment (SandorArpa) [reply src/main/java/uk/gov/pay/connector/refund/dao/RefundDao.java]: My bad, this should've been removed

File: src/main/java/uk/gov/pay/connector/usernotification/service/UserNotificationService.java (hunk 12 above)
Comment (SandorArpa) [reply src/main/java/uk/gov/pay/connector/usernotification/service/UserNotificationService.java:72]: We could just pass in the email and the external_id?

File: src/main/java/uk/gov/pay/connector/tasks/HistoricalEventEmitter.java [hunk 18]
```diff
@@ -122,7 +128,10 @@ public void processRefundEvents(ChargeEntity charge) {
 
     public void emitAndPersistEventForRefundHistoryEntry(RefundHistory refundHistory) {
         Class refundEventClass = RefundStateEventMap.calculateRefundEventClass(refundHistory.getUserExternalId(), refundHistory.getStatus());
-        Event event = EventFactory.createRefundEvent(refundHistory, refundEventClass);
+        ChargeEntity chargeEntity = chargeDao.findByExternalId(refundHistory.getChargeExternalId())
```
Comment (kbottla) [inline src/main/java/uk/gov/pay/connector/tasks/HistoricalEventEmitter.java:131]: Could use `ChargeService` but could be done in PP-6089

File: src/test/java/uk/gov/pay/connector/gateway/stripe/request/StripeTransferInRequestTest.java [hunk 19]
```diff
@@ -33,15 +35,19 @@
     private StripeTransferInRequest stripeTransferInRequest;
     
     @Mock
-    RefundEntity refund;
+    RefundEntity refundEntity;
+//    @Mock
```
Comment (kbottla) [inline src/test/java/uk/gov/pay/connector/gateway/stripe/request/StripeTransferInRequestTest.java]: Commented code can be removed

**Review by kbottla — DISMISSED** [review]
Looks good to me. A couple of minor comments

**Review by kbottla — APPROVED** [review]
